### ISSUE TITLE
variable_name rule vs. `for in`

### ISSUE THREAD

**Issue by brendanjerwin** [realm/SwiftLint#680]
It seems this: `for v in someList {` doesn't trigger the variable_name rule. I would think it should.

**Comment by garnett** [comment]
@jpsim I've tried to fixit and discovered the following:
1. It also happens with other loops, if, guard (any variable in control statement), because
2. `ASTRule` always omits substructures of control statements:

```
guard let subDict = subItem as? [String: SourceKitRepresentable],
                kindString = subDict["key.kind"] as? String,
                kind = KindType(rawValue: kindString) else {
                    return []
            }
```
for example, in provided example `kind` is `source.lang.swift.stmt.forEach` and guard will return early.

I'm interested in fixing that but I honestly don't know where I should start.
For me it looks like a bug of `ASTRule` in general sense and first step is to extend it to "peek" into control structures as well during parsing.

**Comment by norio-nomura** [comment]
`ASTRule.violationRangesInFile` is generic by associatedtype, and customizable which kind the rule need.
e.g.
https://github.com/realm/SwiftLint/pull/826/files#diff-e99fff49b5da9aa0b975161597b27344R64

**Comment by norio-nomura** [comment]
Oops! I pointed wrong function on last comment.
I edited. 🙇

**Comment by marcelofabri** [comment]
This is actually a duplicate of #136. Once https://github.com/apple/swift/pull/11262, we'll get it for free as `source.lang.swift.decl.var.local` declarations.
